### PR TITLE
feat(design-system/Modal): Disclosure

### PR DESCRIPTION
**Pull request by sgangnant**
**What is the problem this PR is trying to solve?**
DS's Modal should support a disclosure prop.

**What is the chosen solution to this problem?**
Implement it.

**Please check if the PR fulfills these requirements**

- [ ] The PR have used `yarn changeset` to a request a release from the CI if wanted.
- [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features) And [non reg](./screenshots.md) done before need review
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: packages/design-system/src/components/Modal/Modal.tsx [hunk 1]
```diff
@@ -30,22 +30,27 @@ export type ModalPropsType = {
...
-		(ref.current as unknown as HTMLElement).focus();
+		if (ref.current) {
+			(ref.current as unknown as HTMLElement).focus();
+		}
 	}, []);
-	const dialog = useDialogState({ visible: true });
 
 	const hasAction = primaryAction || secondaryAction;
 	const onCloseLabel = hasAction ? i18n.t('CLOSE', 'Close') : i18n.t('CANCEL', 'Cancel');
+	const onCloseHandler = disclosure ? () => dialog.setVisible(false) : onClose;
```
Comment (sgangnant) [inline packages/design-system/src/components/Modal/Modal.tsx:53]: Interesting change ☝️

File: packages/design-system/src/components/Modal/Modal.tsx [hunk 2]
```diff
@@ -64,53 +69,61 @@ function Modal(props: ModalPropsType): ReactElement {
...
-									className={styles['modal-header-text__description']}
-									data-test="modal.header.description"
-								>
-									{header.description}
-								</span>
-							)}
-						</div>
-					</div>
+		<>
+			{disclosure && disclosure(() => dialog.setVisible(true))}
```
Comment (sgangnant) [inline packages/design-system/src/components/Modal/Modal.tsx:73]: Interesting change ☝️

Comment (sgangnant) [reply packages/design-system/src/components/Modal/Modal.tsx:73]: Mostly indent 👇

Comment (Benrajalu) [reply packages/design-system/src/components/Modal/Modal.tsx:73]: There's more to the disclosure than the function though: reakit will spread attributes on the disclosure to let users know that it's connected to a modal.

File: packages/design-system/src/components/Modal/Modal.tsx [hunk 3]
```diff
@@ -30,22 +30,27 @@ export type ModalPropsType = {
...
 	preventEscaping?: boolean;
 	children: ReactNode | ReactNode[];
 };
 
 function Modal(props: ModalPropsType): ReactElement {
-	const { header, onClose, primaryAction, secondaryAction, preventEscaping, children } = props;
+	const { header, onClose, disclosure, primaryAction, secondaryAction, preventEscaping, children } =
+		props;
+	const dialog = useDialogState({ visible: !disclosure });
 	const ref = useRef(null);
```
Comment (Benrajalu) [inline packages/design-system/src/components/Modal/Modal.tsx:44]: We shouldn't need this, that should be part of Reakit's job

File: packages/design-system/src/components/Modal/Modal.scss [hunk 4]
```diff
@@ -3,26 +3,33 @@
 $modal-backdrop-z-index: 100000;
 $modal-z-index: $modal-backdrop-z-index + 1;
 
+.modal-backdrop:before,
+.modal-centering {
```
Comment (Benrajalu) [inline packages/design-system/src/components/Modal/Modal.scss:7]:
```suggestion
.modal-container {
```

File: packages/design-system/src/components/Modal/Modal.stories.tsx [hunk 5]
```diff
@@ -140,6 +141,21 @@ export const Basic: ComponentStory<typeof Modal> = props => (
 	</ModalStory>
 );
 
+export const WithDisclosure: ComponentStory<typeof Modal> = props => (
+	<Modal
+		{...props}
+		header={{ title: 'With disclosure' }}
+		onClose={() => {}}
+		disclosure={({ setVisible }: { setVisible: Function }) => (
+			<ButtonPrimary data-test="modal-disclosure" onClick={() => setVisible(true)}>
```
[the remark this answers is not in the document]
Comment (Benrajalu) [reply packages/design-system/src/components/Modal/Modal.stories.tsx:150]: In this story the modal should close on escape, yet it does not 🤔

File: packages/design-system/src/components/Modal/Modal.tsx (hunk 3 above)
Comment (sgangnant) [reply packages/design-system/src/components/Modal/Modal.tsx:44]: Are you're talking about the focus ? It should but it's currently not working https://github.com/ariakit/ariakit/issues/917

File: packages/design-system/src/components/Modal/Modal.stories.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (Benrajalu) [reply packages/design-system/src/components/Modal/Modal.stories.tsx]: Not fixed on my end!

[the remark this answers is not in the document]
Comment (sgangnant) [reply packages/design-system/src/components/Modal/Modal.stories.tsx]: 🤔 can you tell me more about where/how you test it ? ESC works on every story on my side (canvas/docs) except for the one that's not supposed to close on ESC

[the remark this answers is not in the document]
Comment (sgangnant) [reply packages/design-system/src/components/Modal/Modal.stories.tsx]: Indeed, I'll have a look (still not reproducing locally though)

[the remark this answers is not in the document]
Comment (sgangnant) [reply packages/design-system/src/components/Modal/Modal.stories.tsx]: Hmm maybe deployment somehow failed last friday, I just updated the branch and it now works well, could you give it another go on your end please ?
escape

[the remark this answers is not in the document]
Comment (Benrajalu) [reply packages/design-system/src/components/Modal/Modal.stories.tsx]: Much better indeed

**Review by Benrajalu — APPROVED** [review]

**Pull request merged**